### DISCUSSION TITLE
Why Would This Resolve as a BinaryExpression in Validation?

### DISCUSSION THREAD

**Discussion by ballcoach12** [eclipse-langium/langium#1259] [Q&A]
Given the grammar and example shown [here,](https://langium.org/playground?grammar=OYJwhgthYgBAQgGwK4FMBm4KoCIBUBlAKCNQDsAXEAT1hwoGcAFMAYwGsxhUAuI2AbCoBLYNxAMA1AF48IUeICCFEQCNkFVACp%2BggOQAHNp257YZSKmkBJHLD0BvPbsGDUiVNkpTp9Zsa5UAFEPLwodVz0AXz0AbhI-Fg5AkM9yCj5XAipkVgocVFZEGDAKYQB7MlgAH1gXWAAxZDI8irICopKyypr6nFKwAnLkEFZcQuLwbrJ4on6KQeHR8c6ptsz9ABMBhiWxswtsGztHZ1dXbAhVVAkZecWRsYBZT2uQSXro51n7ocfUF5XG4bAS-PaoJotaY1WDZEC5fITLptfg-AZ-ZaQ1qVDbkTYGcrCSjSIJkfGEyjKNQaVC6bAUAAW5U20hejOZAkOVls9gAFGZeUYsKhNLdpCxhaLYPyADRmIWWUU%2BCWKm4ASi0aoA-PY1QK9DwzBRqAYrLy8CbUAAlDA3chjGEAVTIbQtprV2vssTOs2yYAYDL9mjCIPsDAWAb0en5sHYqGo0gIeCt1gAcgBxexygQANzAKCsQQAHgYQKgGAw2gI-QHneH-QzFCBgAwdXo9aiSHWIwzi6Xy5WcZ9mvXI9GzHGE0mUxnpXps8bTbJLTb0HaWqhPTWGd2G02W229XVZpguGFYT3dwH9ww%2BLzZVHNqpDbAn4nk2nM2qYQ%2B9E%2BX2%2B06flmZh6Kw5SIB42JkC%2B4GQYU0zvjOX71NUv5wVB0ywRBmFtEhn7frU6E4Qh6xmBhpGVPhs7zo%2Bz5-qo1Hph69Q-IwqQhnU5ywH2ZYVm0SDlBwqGwNY6BBmklBcYItQAOoMsIHgSeeIkNOUIDKekInbppUlxGcMmwAAajAwhgKoHgdJMpRVvp0kCLUTDyJQukUF6ZgiUwyAWcIAZYtC%2BkiTaFAjGQrnufZMK8QOtnekQsxPMgiBlG6wShOkoYCCZ8jmZZSJrJUqnNNBVnIoVri1A0wjuJsgJvCJYL-P5KIVbAKr0jcrEkGJrk8JFejCOgoExuBZLCIh0X8T07ZmBZQnsCSJZ8YOZCCcJrj8u4DCoGYW2oGtC2TStB3avFJDyYpqC9Z8ADuCkeMN5GVJs414UdtlHnNHCLf2U2rYg81nUQakaQswYZZ86DqY9sDgc0orSNlZkWSs1nTG23qw89r1Ue9lQY7tRaFBob1LTF%2BO6rNAPfXj-2A7MTlEhQvX2KWTNmHmBY-ctbSzMFoUs3oZYhSAZBmMLoUmVztOzLTB19ac9TuJJjB3Ox6WUDoXyzAAUrsZAAPKqAAVghfCK64eZinrlRI7lqNlVUD6wFbPg22Qdso6VBVkBqRBfEDnt5asNlDpEVsHJYxxzi%2Bi5mqlq7rg6tTOq6loejqvL%2BpWwBkGEWrSHohcu-maDc%2BTvtakDRDu0HDs%2B6GXLR0XHOl4WZN-bLHcraGij8bnYSzH3Od5%2BksBEmuEg8d36z1IomwvdCvIOBPNwIESMDULTAB0HjoBQ0isCMZaUFEsDlKaUzqdI-KF9%2BsDyMADIH-Pi9tBqQ8LzjVSr1PtOholZKwgDCIGEKwUOTsV5kEnuvCwNAd57wPkfEAJ8KBnwvjcUo19%2BSSDMLUPQABadsD9RDP1ZElMoICwEQI-iQQBlDQHgOhL-Bg09fo93qAAGXKMAahiBpRQJgfADe8CZ6VF3hgJBx90joMvlgkAN89BaDwfYAA9MQx%2BZDuG8PAYgWhRBtF8PHtAm4rD-71AAMLlAgEKXyPRl6-1gZvBBkjD7SNPufORFBsF6DAGSFRehobfk0QfKxNjTL630WE2x%2BtjGTzMWImC9Q6o3AsfmfhDiTFwGEXAreiSJH7zcSgmRnjMHeIUbyTheAYRVKCDCdM1TagNLqfg6Qxd8EAEI74kKfgfFJIA0mQX0f0wZ-CWFsJ5mHQQjNoA0HqJkmBIz0nb37DmCoyAGBFNQWfeorhVGqNgP02G6TYA3XGgyWA6BirQl0YgXZgheQACJt6PNgMrMI0gADaqZLCbA4hlWwABdaU9yBCoBLIw8aBsvFtFGQXccoKQXcW4jAYAyAwg%2BFpnObMqL0XpExYkv2yLYCnWJTNT0iL9mwAsQyMARJUCbEudcqstzEVEUReCqhrAoUwsqHCwu-JEW8kRa4XFGKZBYudmK-FErCURGRaS5F5K1T1H0amVAN1rBkHrBuTIegyAarMEzG46A2BWG%2Bb8-5lAeBAvmUSV6%2BYpZl3dkbU2eRTpDxQWAagWqHWgIAF7Ak%2BJ81uBYCXsKrM7TmaBw2TMrvYQF3wSAzM3nE0xEyK6hnHBmv6lMYSRSbOAH1Loyj5mEIGuA1R6hwgRAUdA9roS1EiuqzV2qFgbl6K4F1JsEKdsEM6ZxiS%2B3VhEGQYAWKq2uHgOUCCqA-ETvqKmdFbwF2uCvL2Idk7BANDnSLVAozh2wELd6q1FBFCsDGBWdSh6ChrMvc1PlJyt0CCsWQA10EGAPrIAe591KyybHSGZRAn7mWPsgoe9MIov2jNmDu0oIx90nPGeY1wgi15LMglEeZ7z0hfJ%2BdgP5GsMhAphDh4kFqCOnsNIyXyehgW1DIwfCjDKqN6AYMgS%2BdGVV7IOdC6Y%2BYmVQhZScsA%2B815lknvaIkwBYB%2BLeUR7DEKwE8rKbC9J8LBUbURdK7wsqI32NorJ5seLdPlz%2BkSwQiqBDkq1J6otp7z2Xt2HAZDiS%2Br1DQ3AY91BHMDnUjs1wjG8OWqI31EjegQ3GIA0WaQqZHRPHgEEK0Cak1EAHaI-TSTBAYKvhUvQHSAlEJhHoXB35o3t0y0DL93sIGhlKNSTQPgqTyHUJoeV0p0g0HhZ16g7Y2xXKE5USORweTZsFCUDqYp2oijXs7BUk3lQTZmyADUnoZoxz0HHG%2BCdbQn2TrAVOlRUoZ1gKoZkCY5bU3YAzJboo%2BpN1G7HS022Vy7ftKgJ0Lojvp19DkPINWsKRXqy1mkTWVAg7a58cM8I8jDe5CcJwiLLhvB8DWvIRznawGR6YmQaOKD9Is9ZmIQ9wfCFa7weZLBGTNbJzSGEcgFA3Bp%2BT7jggbvU9J%2BTrNkXDClAZKBF8RhGTSEdFaThIF7DPwoAYNkTJNhRhfPSOXrIRRy7o3EEgpJyRM2ZzSbnoE8QEnZvYQXfORdi5S11IgBBKJJL0Oo6U41PA%2BGsMGDU0p%2BQO6dxAF3bvNQ6lmK7zwfB2oAH0bfAHPI5Pn4fUCR-SDdrAsf49SXGe1CPIZ-ZOFgPN5uxPk0x4z2PNPfOi-Wt0NtFPB9bCzFF5w0M20P0yFhLbyQswGdiCZ5zvX8yLEgCO4-JQ3fNAwh%2BYwDvQ%2BGublmH3gfjOQC680PrvQeBhDYBABPm4UYYwPZONmVg-eyDSFn%2B0DAlN1exFgFS7xmxyh9RwEEBoaY6l6Gg%2Bkr56YAaqHzF%2BwF1RPncK6K-45795rIvRjpCAMgfaooMAkAn51pZr8hMgjCbJxYJZJYS4QBEig6xbxaJbJaGaN7PSoF4EYEzSzBj4MCb4L7D4U6RAhp6D3DUHb76pRyjYLgXxgJMQzQX7JreSgJ%2BSgY-xZKsJaqjqVisBfqhieZtT8G%2BQMhfpYaRAGByFjgxjeIGBcFAQ0TZjlZmYrSUyzCQYUBv7gZZrcAUAsFC4MhMQS5YxvrH6VDvoBR6hGFQZCFppTxiHOQSFSEuAyHGGKEuB6AWFWFm7aGZiGb2FH6vrOEfQa5zCoB3rlimGIDuaRC3pgLlgsEACOaANA0gAAivkdQJwr5JYXqPyNvE9G%2BrbpsrEXUakdqDPk4Y0UIaMukfoA0R%2BrkSUUUSUWUeGDNFUeRP%2BoBvmPUWMZQEBiBoNt%2Buks0XAVMaWsBqkVmn3gyuMcBo9HkTcAmMUXsYMRUQkd0VhPUAABIqAGCnFVi1CXHS4MA3GVCzD3HXGtHQSeGsKySoCqBPFZaCAyGvF-FKH6BS4GAsGvhJFZHWAsgRGGEkCvGPHvHMIiGwDfG-HImzyoaOKInAmfBgkMAwwAbJEwm2EzRAyy4cgSyiysLQ7Sahj8ihHFYqGWHMnlBDHFYAYeCaDthAx451oNpViuaZadHWaI7ErY5ih44Y5RGSmo5-b46vDqjtb2AxB8kKn9J9RVQ1RHK1D8kYCCnPEkDamIC1RKkgCNxsEnBPZLjmivaSYdopxfZkDHZAyHYumWiilCCWg%2BA7YOkOj8jVBGg%2BkyB%2BlJybg6Dt72nhl9QSbhlfL6khxYQkZZxeoJjeZWRWYwhsxYFlA5hmiPJ0ljqvK1CPLGriAlmwCPK378GoCVmPJnazp%2BL1k5iEibCPLfiplFrSAZkTBZlAy9lFD67HJGDcrGi4HoFWhtiJpAxxywD4YsZEawDSCNDVSmm6mwC1ymT2wA63FtS3ZryVRCG7k9C1CNTLAnlVC1B1yXkwjnljBfowiJloy8wkAHE0BHGhje4%2BDvl5KZbYpGrBg-klG0zu5Ay-kob6DZ7WAMDhRfDFbZ7RIRKVBwUxAwgGwgC0xHHoWgylAqzYVAwYWuSfmfDZ4RaAXO4yC8gwXhS1BIXyD6yuSdmGbflUU0Vgwqwwj0V2JhQcVhDfju7q6qmpYYUkX0EUU%2B4yBlGigCbOysXSDSWYL8KCUzmzAYVYXlHDnyXqWJJB4QAAXjxAUyA6WZZ6Xu6W5qWYW6XBhZrQWwV8VjzwX4JODcWMUOV6RoVNomVxrYUwjsV4VhC%2BVAz%2BXgzl4bS8hkDlAUDwoFaeiIKOEoAQBkALnfgGjiykIHwQWJJWgZXBX2UBXpBiVQUSUu75WhVuRyVGU2BlUqyCVqmzBZWZY5W9IginCiQ1XnhOWwCiXlEOSwCKXgD8LOVmCuUoXuVuRfC1DeUVzYWqJaC1CL7ljEXlFaCqJAyjW8UFVhWCDxUbXQplLXom7pW9KOHhIMVjVbUUDNXPzrXWIxIXXlXXUZCRQDUCZTUgBHFAyvVpGLrLo3AToXjOTjqboIAzoeDzqJK3WJXJWWChhwTopH62C3VnU8X7W5ahgXzSANIwiY3NI40GAKWNKeKE0tLE2cLWAADSpNmN5NVNqYBsFiigBAQQ1cS6QIVl-5fUehaB%2BB8QcI0mWKXNbcTE8Q06TZZAgtJcBY8KOQO0xWpqwGO0swooWBFg-CPNGB1J2q5gf1FpV%2BnyAADAQgAJyAqSC8gAA628htJtZt2oqi8QKtG8-CtNQQ9NjNzNCsoCcYkV4C20GuTtat-VlNQQXtwgcYAdNwqtAmzSCsAAfIXI7VHc7bAA0vHZHSANHS7XgKHfYAADyJ1ECB0CZVIKx50Z1Z2wgfgZh9SqKPKfIAB6jygKWgjyDtRdydQdtgtdnyoeYABC-qigBCAAWoCp8hbTdKHi3e3QpAvOQEIJ3QJrJAQLXRbVIDPcIHPVUMXfwk8JwqHhYgbE8E8G7XgKvaohbVoOPQwBbQQC3VqJfRbfsvELPQBtvYvfwgQPvYfcfafefU-Y3RbWQBbSANPbEEAA&content=AQBwhgxg1mDmCmwBCB7ARsA3gKGH-Bhw8AdgC4BOAnsAGYCuJEZAliicADJgXUAUASiy4ioogDcewAB4AuAM6UWJWAG0AusAC8wVQCIwegDTAAjOoDcIsUQC%2B1m7aA) the variable declaration (`var x:string[] = ["a", 1];`) is presented as a `BinaryExpression` to my validation routines, and subsequently as a `VariableDeclaration` to my validation routines.  But in the AST it is a `VariableDeclaration`, which is correct.  

I can maybe see why it might look like a `BinaryExpression` given the array initializer, but I cannot figure out why it would appear as both to validation.  I've checked my dependency injection code to make sure that I'm mapping the correct AstNode to the correct validation routine, and Chevrotain is not complaining about ambiguous rules or paths. 

I'm not sure what is going on here.

**Comment by msujew** [comment]
> Chevrotain is not complaining about ambiguous rules or paths

Actually chevrotain is complaining about your rule. See playground console log:

```
Ambiguous Alternatives Detected: <0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1> in <OR1> inside <VariableDeclaration​> Rule,
<string, [, ], =, [> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.
```

In particular the `type=(TypeReference | UnionType))?` alternative is ambiguous. See [here](https://github.com/eclipse-langium/langium/blob/db63283dcc5586107705626efba3b31b428a8d67/packages/langium/src/grammar/langium-grammar.langium#L29-L30) on how to write a union type rule using actions in a non-ambiguous way.

I've just tried reproducing the issue on the latest version of Langium (2.1.0), but on my end the validation for a `BinaryExpression` is never triggered.

**Reply by ballcoach12** [reply]
Thanks so much, Mark! I totally missed the Chevrotain error in the playground.  The union types are a new feature in my language and I haven't been too happy so far with how my implementation works.  I'll definitely look at the link!

**Reply by ballcoach12** [reply]
> I've just tried reproducing the issue on the latest version of Langium (2.1.0), but on my end the validation for a BinaryExpression is never triggered.

I see this behavior as well. I have two unit tests that validate the exact same concrete syntax. In one case it shows up as a BinaryExpression and also a VariableDeclaration; in the other, it only shows up as a VariableDeclaration.

**Reply by spoenemann** [reply]
We should fix this to improve Playground usability: https://github.com/eclipse-langium/langium-website/issues/130
